plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMetadata.java

```diff
@@ -171,6 +175,33 @@ private JdbcTableHandle flushAttributesAsQuery(ConnectorSession session, JdbcTab
                 handle.getNextSyntheticColumnId());
     }
 
+    private JdbcTableHandle flushTopNAttributesAsQuery(ConnectorSession session, JdbcTableHandle handle, List<SortItem> sortItems, long limit)
+    {
+        verify(!sortItems.isEmpty(), "sortItems is empty");
+        return transformQueryRelationPreparedQuery(session, handle, preparedQuery ->
+                jdbcClient.prepareTopNQuery(preparedQuery, sortItems, limit));
+    }
+
+    private JdbcTableHandle transformQueryRelationPreparedQuery(ConnectorSession session, JdbcTableHandle handle, Function<PreparedQuery, PreparedQuery> transform)
+    {
+        List<JdbcColumnHandle> columns = jdbcClient.getColumns(session, handle);
+
+        if (!(handle.getRelationHandle() instanceof JdbcQueryRelationHandle)) {
+            handle = flushAttributesAsQuery(session, handle);
+        }
+
+        checkState(handle.getRelationHandle() instanceof JdbcQueryRelationHandle,
+                "tableHandle.getRelationHandle() is not JdbcQueryRelationHandle after flushing attributes");
```

Comment:
The actually more useful conditions are

- handle has no limit
- handle has no contraints

-- because you're operating on PreparedQuery directly, without applying these.
Note that it wasn't the case in `applyAggregation`, because that method was not ignoring a limit or constraint, should it have survived `flushAttributesAsQuery` -- because it was using the same `prepareQuery` method, which it shouldn't anymore, but that's out of scope of this PR.